### PR TITLE
Add formatter support for call and class definition `Arguments`

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

This PR leverages the `Arguments` AST node introduced in #6259 in the formatter, which ensures that we correctly handle trailing comments in calls, like:

```python
f(
  1,
  # comment
)

pass
```

(Previously, this was treated as a leading comment on `pass`.)

This also allows us to unify the argument handling across calls and class definitions.

## Test Plan

A bunch of new fixture tests, plus improved Black compatibility.

### CODE REVIEW

File: crates/ruff_python_formatter/src/builders.rs [hunk 1]
```diff
@@ -252,7 +252,7 @@ impl<'ast> Format<PyFormatContext<'ast>> for EmptyWithDanglingComments<'_> {
             [group(&format_args![
                 self.opening,
                 // end-of-line comments
-                dangling_comments(&self.comments[..end_of_line_split]),
+                trailing_comments(&self.comments[..end_of_line_split]),
```
Comment (charliermarsh) [inline crates/ruff_python_formatter/src/builders.rs:255]: We don't want these to break, IIUC.

File: crates/ruff_python_formatter/src/statement/stmt_class_def.rs [hunk 2]
```diff
@@ -71,58 +87,3 @@ impl FormatNodeRule<StmtClassDef> for FormatStmtClassDef {
         Ok(())
     }
 }
-
-struct FormatInheritanceClause<'a> {
-    class_definition: &'a StmtClassDef,
-}
-
-impl Format<PyFormatContext<'_>> for FormatInheritanceClause<'_> {
```
Comment (charliermarsh) [inline crates/ruff_python_formatter/src/statement/stmt_class_def.rs:79]: We get all of this for free now, I think.

File: crates/ruff_python_formatter/src/other/arguments.rs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+use crate::expression::parentheses::{
+    parenthesized, NeedsParentheses, OptionalParentheses, Parentheses,
+};
+use crate::prelude::*;
+use crate::FormatNodeRule;
+
+#[derive(Default)]
+pub struct FormatArguments;
+
+impl FormatNodeRule<Arguments> for FormatArguments {
```
Comment (charliermarsh) [inline crates/ruff_python_formatter/src/other/arguments.rs:19]: This is largely moved from `expr_call`.

**Review by zanieb — APPROVED** [review]

**Comment by charliermarsh** [comment]
Looking into ecosystem failure.

File: crates/ruff_python_formatter/src/comments/placement.rs [hunk 4]
```diff
@@ -843,6 +847,32 @@ fn handle_leading_function_with_decorators_comment(comment: DecoratedComment) ->
...
+/// ```python
+/// @dataclass
+/// # comment
+/// class Foo(Bar):
+///     ...
+/// ```
+///
+/// The comment should be attached to the enclosing [`ast::StmtClassDef`] as a dangling node,
+/// as opposed to being treated as a leading comment on `Bar` or similar.
+fn handle_leading_class_with_decorators_comment<'a>(
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/comments/placement.rs:862]: Nit: This seems unrelated to this PR.

File: crates/ruff_python_formatter/src/other/arguments.rs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+impl NeedsParentheses for ExprCall {
+    fn needs_parentheses(
+        &self,
+        _parent: AnyNodeRef,
+        context: &PyFormatContext,
+    ) -> OptionalParentheses {
+        self.func.needs_parentheses(self.into(), context)
+    }
+}
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/other/arguments.rs:127]: This should not have been moved. It belongs to `ExprCall`

File: crates/ruff_python_formatter/src/statement/stmt_class_def.rs [hunk 6]
```diff
@@ -23,40 +20,84 @@ impl FormatNodeRule<StmtClassDef> for FormatStmtClassDef {
...
-        if !decorator_list.is_empty() {
-            hard_line_break().fmt(f)?;
+            if leading_definition_comments.is_empty() {
+                write!(f, [hard_line_break()])?;
+            } else {
+                // Write any leading definition comments (between last decorator and the header)
+                // while maintaining the right amount of empty lines between the comment
+                // and the last decorator.
+                let decorator_end =
+                    skip_trailing_trivia(last_decorator.end(), f.context().source());
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/statement/stmt_class_def.rs:44]: Nit: I find this name confusing. I would expect it to take an iterator as input (or something stateful that is after positioned after the trailing trivia). How about `find_first_non_trailing_trivia_position` or `first_non_trailing_trivia`

Related: There's a `token.kind().is_triva` method to avoid listing all trivia kinds.

File: crates/ruff_python_formatter/src/statement/stmt_class_def.rs [hunk 7]
```diff
@@ -23,40 +20,84 @@ impl FormatNodeRule<StmtClassDef> for FormatStmtClassDef {
...
+            // If the arguments contain any dangling own-line comments, we retain the parentheses,
+            // e.g., in:
+            // ```python
+            // class A(  # comment
+            //     # comment
+            // ):
+            //     ...
+            // ```
+            if arguments.args.is_empty()
+                && arguments.keywords.is_empty()
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/statement/stmt_class_def.rs:83]: Nit: Add a `arguments.is_empty()` helper.

File: crates/ruff_python_formatter/src/statement/stmt_class_def.rs [hunk 8]
```diff
@@ -23,40 +20,84 @@ impl FormatNodeRule<StmtClassDef> for FormatStmtClassDef {
...
-        {
-            parenthesized(
-                "(",
-                &FormatInheritanceClause {
-                    class_definition: item,
-                },
-                ")",
-            )
-            .fmt(f)?;
+        if let Some(arguments) = arguments {
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/statement/stmt_class_def.rs:61]: Same: I think this change should have been a separate PR because it goes beyond the refactor.

File: crates/ruff_python_formatter/src/statement/stmt_class_def.rs [hunk 9]
```diff
@@ -23,40 +20,84 @@ impl FormatNodeRule<StmtClassDef> for FormatStmtClassDef {
...
+            // ```
+            if arguments.args.is_empty()
+                && arguments.keywords.is_empty()
+                && comments
+                    .dangling_comments(arguments)
+                    .iter()
+                    .all(|comment| comment.line_position().is_end_of_line())
+            {
+                let dangling = comments.dangling_comments(arguments);
+                write!(f, [trailing_comments(dangling)])?;
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/statement/stmt_class_def.rs:90]: I think this is the first usage where another node formats a node's dangling comments. I also don't understand the reason for it from the comment above because all examples in the comment above have parentheses, but the point here seems to be to remove the parentheses. 

Why not move this into `Arguments`? It removes the need for querying the dangling comments here (we only have to do it in arguments) and is more in line of what we do elsewhere.

File: crates/ruff_python_formatter/src/comments/placement.rs [hunk 10]
```diff
@@ -1105,6 +1135,64 @@ fn find_only_token_in_range(
...
+            ([], [first_keyword, ..]) => Some(first_keyword.range()),
+            ([], []) => None,
+        };
+
+        if let Some(first_argument) = first_argument {
+            if arguments.start() < comment.start() && comment.end() < first_argument.start() {
+                return CommentPlacement::dangling(comment.enclosing_node(), comment);
+            }
+        }
+    }
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/comments/placement.rs:1191]: Nit: A more generalized implementation of this could be:

```rust
if comment.line_position().is_end_of_line() {
    let mut lexer = SimpleTokenizer::new(
        locator.contents(),
        TextRange::new(arguments.start(), comment.start()),
    )
    .skip_trivia()
    .skip_while(|t| {
        matches!(
            t.kind(),
            SimpleTokenKind::LParen | SimpleTokenKind::LBrace | SimpleTokenKind::LBracket
        )
    });

    if lexer.next().is_none() {
        return CommentPlacement::dangling(arguments, comment);
    }
}

CommentPlacement::Default(comment)
```

File: crates/ruff_python_formatter/src/statement/stmt_class_def.rs (hunk 8 above)
Comment (charliermarsh) [reply crates/ruff_python_formatter/src/statement/stmt_class_def.rs:61]: To me this is squarely part of the PR and the desired refactor.

File: crates/ruff_python_formatter/src/statement/stmt_class_def.rs (hunk 7 above)
Comment (charliermarsh) [reply crates/ruff_python_formatter/src/statement/stmt_class_def.rs:83]: Good call, I believe this now exists on main but didn't at the time.

File: crates/ruff_python_formatter/src/statement/stmt_class_def.rs (hunk 9 above)
Comment (charliermarsh) [reply crates/ruff_python_formatter/src/statement/stmt_class_def.rs:90]: I could handle it in placement instead of here?

Comment (MichaReiser) [reply crates/ruff_python_formatter/src/statement/stmt_class_def.rs:90]: That might be easier to understand if it is easy enough

File: crates/ruff_python_formatter/src/comments/placement.rs (hunk 10 above)
Comment (zanieb) [reply crates/ruff_python_formatter/src/comments/placement.rs:1191]: cc @charliermarsh are we going to use this to consolidate the Arguements / TypeParams implementations?

Comment (charliermarsh) [reply crates/ruff_python_formatter/src/comments/placement.rs:1191]: I can do another pass on this and consolidate.

File: crates/ruff_python_formatter/src/statement/stmt_class_def.rs (hunk 6 above)
Comment (charliermarsh) [reply crates/ruff_python_formatter/src/statement/stmt_class_def.rs:44]: What was "Related 2" here?

File: crates/ruff_python_formatter/src/comments/placement.rs (hunk 10 above)
Comment (charliermarsh) [reply crates/ruff_python_formatter/src/comments/placement.rs:1191]: https://github.com/astral-sh/ruff/pull/6315

File: crates/ruff_python_formatter/src/statement/stmt_class_def.rs (hunk 6 above)
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/statement/stmt_class_def.rs:44]: Whoops. That I was unsure if it should return an option

Comment (charliermarsh) [reply crates/ruff_python_formatter/src/statement/stmt_class_def.rs:44]: :joy:
